balancer_conn_wrappers.go

```diff
@@ -160,35 +195,133 @@ func (ccb *ccBalancerWrapper) handleSubConnStateChange(sc balancer.SubConn, s co
+func (ccb *ccBalancerWrapper) handleSubConnStateChange(update *scStateUpdate) {
 	ccb.balancerMu.Lock()
-	defer ccb.balancerMu.Unlock()
-	return ccb.balancer.UpdateClientConnState(*ccs)
+	ccb.balancer.UpdateSubConnState(update.sc, balancer.SubConnState{ConnectivityState: update.state, ConnectionError: update.err})
+	ccb.balancerMu.Unlock()
+}
+
+func (ccb *ccBalancerWrapper) exitIdle() {
+	if ccb == nil || ccb.cc.GetState() != connectivity.Idle {
+		return
+	}
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeExitIdle})
+}
+
+func (ccb *ccBalancerWrapper) handleExitIdle() {
+	ccb.balancerMu.Lock()
+	ccb.balancer.ExitIdle()
+	ccb.balancerMu.Unlock()
 }
 
 func (ccb *ccBalancerWrapper) resolverError(err error) {
 	if ccb == nil {
 		return
 	}
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeResolverError, update: err})
+}
+
+func (ccb *ccBalancerWrapper) handleResolverError(err error) {
 	ccb.balancerMu.Lock()
-	defer ccb.balancerMu.Unlock()
 	ccb.balancer.ResolverError(err)
+	ccb.balancerMu.Unlock()
+}
+
+// switchTo is invoked by grpc to instruct the balancer wrapper to switch to the
+// LB policy identified by name.
+//
+// ClientConn calls newCCBalancerWrapper() at creation time. Upon receipt of the
+// first good update from the name resolver, it determines the LB policy to use
+// and invokes the switchTo() method. Upon receipt of every subsequent update
+// from the name resolver, it invokes this method.
+//
+// the ccBalancerWrapper keeps track of the current LB policy name, and skips
+// the graceful balancer switching process if the name does not change.
+func (ccb *ccBalancerWrapper) switchTo(name string) {
+	if ccb == nil {
+		return
+	}
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeSwitchTo, update: name})
+}
+
+// handleSwitchTo handles a balancer switch update from the update channel. It
+// calls the SwitchTo() method on the gracefulswitch.Balancer with a
+// balancer.Builder corresponding to name. If no balancer.Builder is registered
+// for the given name, it uses the default LB policy which is "pick_first".
+func (ccb *ccBalancerWrapper) handleSwitchTo(name string) {
+	ccb.balancerMu.Lock()
+	defer ccb.balancerMu.Unlock()
+
+	if strings.EqualFold(ccb.curBalancerName, name) {
+		return
+	}
+
+	channelz.Infof(logger, ccb.cc.channelzID, "ClientConn switching balancer to %q", name)
+	builder := balancer.Get(name)
+	if builder == nil {
+		channelz.Warningf(logger, ccb.cc.channelzID, "Channel switches to new LB policy %q due to fallback from invalid balancer name", PickFirstBalancerName)
+		channelz.Infof(logger, ccb.cc.channelzID, "failed to get balancer builder for: %v, using pick_first instead", name)
+		builder = newPickfirstBuilder()
+	} else {
+		channelz.Infof(logger, ccb.cc.channelzID, "Channel switches to new LB policy %q", name)
+	}
+
+	if err := ccb.balancer.SwitchTo(builder); err != nil {
+		channelz.Errorf(logger, ccb.cc.channelzID, "Channel failed to build new LB policy %q: %v", name, err)
```

Comment:
Is this sufficient?  What happens to RPCs at this point?  I think we should be switching to some kind of erroring LB policy so this error message can be reported via RPCs.